fendermint/vm/topdown/src/sync.rs

```diff
@@ -3,139 +3,245 @@
 //! A constant running process that fetch or listener to parent state
 
 use crate::error::Error;
-use crate::{BlockHash, BlockHeight, Config, InMemoryFinalityProvider, ParentFinalityProvider};
+use crate::{
+    BlockHash, BlockHeight, CachedFinalityProvider, Config, IPCParentFinality,
+    ParentFinalityProvider, Toggle,
+};
 use anyhow::{anyhow, Context};
-use async_stm::atomically_or_err;
+use async_stm::{atomically, atomically_or_err};
 use fvm_shared::clock::ChainEpoch;
 use ipc_agent_sdk::apis::IpcAgentClient;
 use ipc_agent_sdk::jsonrpc::JsonRpcClientImpl;
 use ipc_agent_sdk::message::ipc::ValidatorSet;
 use ipc_sdk::cross::CrossMsg;
 use ipc_sdk::subnet_id::SubnetID;
+use std::cmp::min;
 use std::sync::Arc;
 use std::time::Duration;
 
+/// The max number of blocks polling should query each parent view update. If the number of blocks
+/// polled equals this value, it would stop polling for this iteration and commit the result to cache.
+const MAX_PARENT_VIEW_BLOCK_GAP: BlockHeight = 100;
+/// When polling parent view, if the number of top down messages exceeds this limit,
+/// the polling will stop for this iteration and commit the result to cache.
+const TOPDOWN_MSG_LEN_THRESHOLD: usize = 500;
```

Comment:
It is possible to place it in the config but I think they should be aligned by the protocol. Say if the `MAX_PARENT_VIEW_BLOCK_GAP` is reached, the fetched data would be committed into cache and parent finality would start producing proposals. In a way, these parameters control what the height would be in the proposals, if we use config, then anyone can input some value, I'm not sure how this would impact the consensus. If there are no impact, then in config should work too.

Reply:
My thinking on issues like this is that:
* if something affects consensus, ie. it's is vital that we all use the same value otherwise we might be rejecting proposals, then it should be in the genesis block
* if it's something validators can choose then it should be in the config, because otherwise it's an arbitrary number in the code we cannot easily change when we realise there is a better value

In this case it depends on whether we want these be a basis of rejection. 

Like I commented somewhere I think we won't be able to apply limits directly on these, more like on gas, and even then it's impossible to say up front what's a good value because it depends on what else is in the block, so I recommended that we actually append values into a queue rather than execute the whole checkpoint as one, and then chip away at it deterministically in the upcoming blocks.

So for now I think these are more about protecting the process from out-of-memory problems, and should be in the config.

Reply:
This works for me!